File: Modeling .NET/VB/Queens/Queens/Queen.vb
Imports Simulation

Public Class Queen
    Inherits Coroutine

    Public Row As Byte
    Public Col As Byte

    Public Sub New(ACol as Byte)
        Col = ACol
        Row = 0
    End Sub

    Protected Overrides Sub Run()
        Do
            If Board.IsFree(Col, Row) Then
                Board.MakeOccupied(Col, Row)
                If Col < Board.QueenCount - 1 Then
                    SwitchTo(Board.Queens(Col + 1))
                Else
                    Board.Remember()
                End If
                Board.MakeFree(Col, Row)
            End If
            Row = Row + 1
            If Row = Board.QueenCount Then
                Row = 0
                If Col = 0 Then
                    Exit Sub
                Else
                    SwitchTo(Board.Queens(Col - 1))
                End If
            End If
        Loop
    End Sub
End Class

File: Modeling .NET/Delphi/Queens/Queen.pas
unit Queen;

interface
uses Simulation;

type
  TQueen = class(Coroutine)
  strict protected
    { Private Declarations }
    procedure Run; override;
  public
    Col : Byte;
    Row : Byte;
    constructor Create(ACol : Byte);
  end;

implementation
uses Board;

{ TQueen }

constructor TQueen.Create(ACol : Byte);
begin
  inherited Create;
  Col := ACol;
  Row := 0;
end;

procedure TQueen.Run;
begin
  inherited;
  while True do
  begin
    if TBoard.IsFree(Col, Row) then
    begin
      TBoard.MakeOccupied(Col, Row);
      if Col < QueensCount - 1 then
        TBoard.Queens[Col + 1].SwitchTo
      else
        TBoard.Remember;
      TBoard.MakeFree(Col, Row);
    end;
    Inc(Row);
    if Row = QueensCount then
    begin
      Row := 0;
      if Col = 0 then
        Exit
      else
        TBoard.Queens[Col - 1].SwitchTo;
    end;
  end;
end;

end.
